src/cache/core/cache.ts

```diff
@@ -1,19 +1,52 @@
 import type { DocumentNode } from "graphql";
 import { wrap } from "optimism";
 
-import type { StoreObject, Reference } from "../../utilities/index.js";
+import type {
+  StoreObject,
+  Reference,
+  DeepPartial,
+} from "../../utilities/index.js";
 import {
+  Observable,
   cacheSizes,
   defaultCacheSizes,
   getFragmentQueryDocument,
+  mergeDeepArray,
 } from "../../utilities/index.js";
 import type { DataProxy } from "./types/DataProxy.js";
 import type { Cache } from "./types/Cache.js";
 import { WeakCache } from "@wry/caches";
 import { getApolloCacheMemoryInternals } from "../../utilities/caching/getMemoryInternals.js";
+import type {
+  OperationVariables,
+  TypedDocumentNode,
+} from "../../core/types.js";
+import { equal } from "@wry/equality";
+import type { MissingTree } from "./types/common.js";
 
 export type Transaction<T> = (c: ApolloCache<T>) => void;
 
+// todo: confirm we can start with this limited set of options
+export interface WatchFragmentOptions<TData, TVars> {
+  fragment: DocumentNode | TypedDocumentNode<TData, TVars>;
+  from: StoreObject | Reference | string;
+  fragmentName?: string;
+  optimistic?: boolean;
+}
+
+// todo: refactor duplicated type, see UseFragmentResult
+export type WatchFragmentResult<TData> =
+  | {
+      data: TData;
+      complete: true;
+      missing?: undefined;
+    }
+  | {
+      data: DeepPartial<TData>;
+      complete: false;
+      missing?: MissingTree | undefined;
```

Comment:
Is there ever a case where the result isn't complete and we have an `undefined` missing? If not, I'd make this a required property and remove the `| undefined` part of the union.

```suggestion
      missing: MissingTree;
```

Reply:
Updated in https://github.com/apollographql/apollo-client/pull/11465/commits/67c1e1cb18b242644a36b832607b2abaea4a6714